### PR TITLE
feat: add PartiQL language aserver

### PR DESCRIPTION
**Pull request by akoreman**
## Problem

Adds an LSP server for the [PartiQL language](https://partiql.org/). This server offers diagnostics for PartiQL files by parsing the file using the [Rust parser maintained by the PartiQL team](https://github.com/partiql/partiql-lang-rust).

This parser is compiled to a WebAssembly binary using the bindings maintained by the PartiQL team in the [PartiQL playground package](https://github.com/partiql/partiql-rust-playground). This binary is further encoded as a base64 string in a Javascript file to make it easy to be bundled together with the other assets of the server. See the README for instructions on how to update the binary.

## Solution

- Tested the server wrapped in the standalone runtime in VS Code:

Screenshot 2024-05-20 at 23 54 32

- Tested the server wrapped in the webworker runtime a web-based LSP client.
- Added unit tests covering both the server and the service.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
> This binary is further encoded as a base64 string in a Javascript file to make it easy to be bundled together with the other assets of the server.

Why can't it be bundled as a wasm file?

File: app/aws-lsp-partiql-binary/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { PartiQLLanguageServer } from '@aws/aws-lsp-partiql'
```
Comment (viktorsaws) [inline app/aws-lsp-partiql-binary/src/index.ts:1]: nit: `aws-` prefix in package name is redundant

File: server/aws-lsp-partiql/src/server/language-server.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { type DidChangeTextDocumentParams, type Server } from '@aws/language-server-runtimes/server-interface'
+import { TextDocumentSyncKind, type DidOpenTextDocumentParams } from 'vscode-languageserver'
+import { TextDocument } from 'vscode-languageserver-textdocument'
```
Comment (viktorsaws) [inline server/aws-lsp-partiql/src/server/language-server.ts:3]: nit: `TextDocument` interface is available through Runtimes `@aws/language-server-runtimes/server-interface`, so direct dependency on `vscode-languageserver-textdocument` in code.

`TextDocumentSyncKind` and `DidOpenTextDocumentParams` should be also available, so direct dependency on VSCode LSP packages can be minimized even more.

File: server/aws-lsp-partiql/src/server/language-service.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import type { Diagnostic } from 'vscode-languageserver-protocol'
```
Comment (viktorsaws) [inline server/aws-lsp-partiql/src/server/language-service.ts:1]: nit: imports from `vscode-languageserver-protocol` should be available from  `@aws/language-server-runtimes/server-interface`

**Review by viktorsaws — APPROVED** [review]

File: app/aws-lsp-partiql-binary/src/index.ts (hunk 1 above)
Comment (akoreman) [reply app/aws-lsp-partiql-binary/src/index.ts:1]: yeah that `aws` there is a bit redundant, removed it

**Comment by akoreman** [comment]
@justinmk3 Main reason for not bundling it as a wasm file is to make it easier to use the server as a web worker when assets are loaded from a CDN. If the binary is provided as a wasm file, the worker needs to know which CDN URL to fetch the binary from. This URL needs to be communicated by the client to the web worker which adds complexity to the integration into a web app.

**Review by viktorsaws — APPROVED** [review]

**Pull request merged**